### PR TITLE
Docs for INSERT and UPSERT

### LINKED ISSUE #302: INSERT statement

### PR DESCRIPTION
**Pull request by jseldess**
Fixes #302 and #279 

@paperstreet, PTAL.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/308)

### CODE REVIEW

**Comment by jseldess** [comment]
@paperstreet, here are html version, in case you want to view that way:
- http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/insert.html
- http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/upsert.html

**Comment by paperstreet** [comment]
Great stuff! I've never reviewed public facing docs before, so I erred on the side of overcommunication. Feel free to ignore any suggestions you feel aren't really appropriate or are too pedantic

<details><summary><i>Previously, jseldess wrote…</i></summary>
> @paperstreet, here are html version, in case you want to view that way:
> 
> - http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/insert.html
> 
> - http://cockroach-draft-docs.s3-website-us-east-1.amazonaws.com/upsert.html
</details>


---

Review status: 0 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[insert.md, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDNQUNB8UQBqjd-Ik8:-KIDNQUNB8UQBqjd-Ik9:-210993253) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L48)):*
> ```Markdown
> ### Default Values
> 
> [Defaults values](data-definition.html) are used when you leave specific columns out of your statement, or when you explicitly request default values. For example, both of the following statements would create a row with `balance` filled with its default value, in this case `NULL`:
> ```

nit: Default values

---

*[insert.md, line 51 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDOb_-5BxT0b4uPS0d:-KIDOb_0wxlxn6qMxRko:687036874) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L51)):*
> ```Markdown
> 
> ~~~ sql
> INSERT INTO accounts (id, balance) VALUES (5);
> ```

I'm surprised this works. Seems a little fishy even if it does. I think of the "leave specific columns out of your statement" part more in terms of columns in the table that you've left out of the insert statement: `create table ab (a int, b int default 7); insert into ab (a) values (1);` will use the default for b

tl;dr consider leaving ", balance" out of this line

---

*[insert.md, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDOTZdMHYEbg3Gmz5o:-KIDOTZdMHYEbg3Gmz5p:125494812) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L77)):*
> ```Markdown
> ~~~
> 
> To return all columns of the inserted rows, use the `*` character:
> ```

Dunno how fine of a point you want to put on it, but this returns all columns in the table, not all columns being inserted.

---

*[insert.md, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDPdsn5Uib8MxrviOj:-KIDPdsn5Uib8MxrviOk:-1864214650) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L106)):*
> ```Markdown
> ### Update Values with `ON CONFLICT`
> 
> Normally, when inserted values conflict with a `UNIQUE` constraint on one or more columns, CockroachDB returns an error. To update the affected rows instead, use an `ON CONFLICT` clause containing the column with the unique conflict and the `DO UPDATE SET` expression set to the columns to be updated:
> ```

if the unique index has more than one column, you have to list all of them (and in the same order as in the index)

---

*[insert.md, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDQ5_YUnXMFf0d2xfd:-KIDQ5_Z9x58GhETLEjX:982298318) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L112)):*
> ```Markdown
> ~~~ 
> 
> When a unique conflict is detected, CockroachDB temporarily stores the row(s) in a table called `excluded`. As you can see above, you set the columns to be update to the corresponding columns in the temporary `excluded` table.
> ```

nit: row not row(s)
nit: "to be updated to"

---

*[insert.md, line 116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDQN1Rxiz6WdKzohcF:-KIDQN1SfxuKfNi6K_Rw:1809324180) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L116)):*
> ```Markdown
> Note the following limitations:
> 
> -   It's not possible to update the same row twice with a single `INSERT ... ON CONFLICT` statement. For example, the following would not be allowed:
> ```

missing example?

---

*[insert.md, line 123 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDQYZLDDOE1LXeRT9i:-KIDQYZLDDOE1LXeRT9j:-1711587426) ([raw file](https://github.com/cockroachdb/docs/blob/1df2fd19e8cb1cbfc3c3627ab86a96282b96df79/insert.md#L123)):*
> ```Markdown
> ### Ignore Insert `ON CONFLICT`
> 
> To avoid an error in case an insert conflicts with a `UNIQUE` constraint, set the `ON CONFLICT` clause to `DO NOTHING`:
> ```

"to avoid an error" gives me an incomplete picture. maybe explicitly say that any rows that conflict leave the ones in the database untouched and that the statement succeeds

---

*[insert.md, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDSd4LnrrrEa-2CaK5:-KIDSd4LnrrrEa-2CaK6:1292620310) ([raw file](https://github.com/cockroachdb/docs/blob/b95a4b1e8e2b333dff1bebed63106ad040d524bb/insert.md#L29)):*
> ```Markdown
> `select_stmt` | A [`SELECT`](select.html) statement to retrieve values from another table and insert them as new rows. The target columns in the `SELECT` statement must match the type of the columns being inserted into.
> `opt_on_conflict` |
> `opt_on_conflict` | 
> ```

opt_on_conflict twice?

---

*[insert.md, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDT9q99P7CyN9zwBKK:-KIDT9q99P7CyN9zwBKL:1704819603) ([raw file](https://github.com/cockroachdb/docs/blob/b5ad7e5304dc5125b7753cae2007c1fc0a940536/insert.md#L34)):*
> ```Markdown
> `DEFAULT VALUES` | To fill all columns with their [default values](data-definition.html#default-value), use `DEFAULT VALUES` in place of `select_stmt`. To fill a specific column with its default value, leave the value out of the `select_stmt` or use `DEFAULT` at the appropriate position. See the [Insert Default Values](#insert-default-values) examples below.
> `RETURNING target_list` | Return values based on rows inserted, where `target_list` can be specific column names from the table, `*` for all columns, or a computation on specific columns. See the [Insert and Return Values](#insert-and-return-values) example below. 
> `on_conflict` | Normally, when inserted values conflict with a `UNIQUE` constraint on one or more columns, CockroachDB returns an error. To update the affected rows instead, use an `ON CONFLICT` clause containing the column with the unique constraint and the `DO UPDATE SET` expression set to the column(s) to be updated. To avoid an error, set the `ON CONFLICT` clause to `DO NOTHING`. See the [Update Values ON CONFLICT](#update-values-on-conflict) and [Ignore Insert ON CONFLICT](#ignore-insert-on-conflict) examples below.<br><br>Note that it's not possible to update the same row twice with a single `INSERT ... ON CONFLICT` statement. Also, if the values in the `SET` expression cause uniqueness conflicts, CockroachDB will return an error.<br><br>As a short-hand alternative to the `ON CONFLICT` clause, you can use the [`UPSERT`](upsert.html) statement. However, `UPSERT` does not let you specify the column with the unique constraint; it assumes that the column is the primary key. Using `ON CONFLICT` is therefore more flexible.
> ```

FWIW, the set expressions are also more flexible

---

*[insert.md, line 12 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDVCEmIVQBg2HqMRgD:-KIDVCEnvo8haYqB1Mau:1113522587) ([raw file](https://github.com/cockroachdb/docs/blob/90015851270912ca60aa16b57cdc121a7a4f57c9/insert.md#L12)):*
> ```Markdown
> ## Required Privileges
> 
> The user must have the `INSERT` [privilege](privileges.html) on the table. To use `ON CONFLICT`, the user must also have the `UPDATE` privilege on the table.
> ```

UPDATE should only be necessary for ON CONFLICT DO UPDATE. DO NOTHING should still only require INSERT (though I realize this is not currently true, I'll go file an issue: https://github.com/cockroachdb/cockroach/issues/6826)

---

*[upsert.md, line 8 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDUF67ax2csouha3Cr:-KIDUF68096q2w2vfMno:1927486892) ([raw file](https://github.com/cockroachdb/docs/blob/b5ad7e5304dc5125b7753cae2007c1fc0a940536/upsert.md#L8)):*
> ```Markdown
> The `UPSERT` [statement](sql-statements.html) is short-hand for [`INSERT ... ON CONFLICT`](insert.html). It inserts rows in cases where specified values do not violate uniqueness constraints, and it updates rows in cases where values do violate uniqueness constraints. 
> 
> Note that `UPSERT` does not let you specify the column with the unique constraint; it assumes that the column is the primary key. Using `INSERT ... ON CONFLICT` is therefore more flexible.
> ```

Any UPSERT statement can be exactly transformed into an INSERT ... ON CONFLICT. I'd love to see this transformation called out.

`UPSERT INTO foo (a, b) VALUES (1, 2)` *exactly* transforms into `INSERT INTO foo (a, b) VALUES (1, 2) ON CONFLICT (<primary key columns>) DO UPDATE SET b = excluded b` where there is a SET expr for every column that's in qualified_name list but not in on_conflict

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/308#-:-KIDR4J7aL0EgiaVqGue:1432982226)*

**Comment by jseldess** [comment]
Review status: 0 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[insert.md, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDNQUNB8UQBqjd-Ik8:-KID_4_j4v33C7u-zPQY:235009160) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L48)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> nit: Default values
</details>
Sorry, you're looking at an earlier commit again. I need to learn how to clean them up.

---

*[insert.md, line 51 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDOb_-5BxT0b4uPS0d:-KIDa0-gONifSZpC8zMP:752963001) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L51)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> I'm surprised this works. Seems a little fishy even if it does. I think of the "leave specific columns out of your statement" part more in terms of columns in the table that you've left out of the insert statement: `create table ab (a int, b int default 7); insert into ab (a) values (1);` will use the default for b
> 
> tl;dr consider leaving ", balance" out of this line
</details>
Taking `balance` out of that example.

---

*[insert.md, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDOTZdMHYEbg3Gmz5o:-KIDdHIluF26a8b8ZNUj:1119442703) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L77)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Dunno how fine of a point you want to put on it, but this returns all columns in the table, not all columns being inserted.
</details>
I think the latest version is ok, but let me know if not.

---

*[insert.md, line 106 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDPdsn5Uib8MxrviOj:-KIDdmaxyAtExwk4qGM-:-499146009) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L106)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> if the unique index has more than one column, you have to list all of them (and in the same order as in the index)
</details>
Done.

---

*[insert.md, line 112 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDQ5_YUnXMFf0d2xfd:-KIDdsZ8q50u8EaaW2bv:1060997087) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L112)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> nit: row not row(s)
> nit: "to be updated to"
</details>
I think this is correct in the latest version. Please let me know if you still see trouble.

---

*[insert.md, line 116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDQN1Rxiz6WdKzohcF:-KIDe2EW8aBk1mHPjI2O:556428031) ([raw file](https://github.com/cockroachdb/docs/blob/10f2e83ce7e3c851038109fd8cbbc2f7c03a02ed/insert.md#L116)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> missing example?
</details>
From an earlier commit. Shouldn't be a problem in the latest version.

---

*[insert.md, line 123 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDQYZLDDOE1LXeRT9i:-KIDigS1Vvf1FZ1wQKK0:-690319306) ([raw file](https://github.com/cockroachdb/docs/blob/1df2fd19e8cb1cbfc3c3627ab86a96282b96df79/insert.md#L123)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> "to avoid an error" gives me an incomplete picture. maybe explicitly say that any rows that conflict leave the ones in the database untouched and that the statement succeeds
</details>
Fixed this language a bit and enhanced the related example toward the bottom of the page.

---

*[insert.md, line 29 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDSd4LnrrrEa-2CaK5:-KIDZh22Ukrz-3UtEOo-:-2008113677) ([raw file](https://github.com/cockroachdb/docs/blob/b95a4b1e8e2b333dff1bebed63106ad040d524bb/insert.md#L29)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> opt_on_conflict twice?
</details>
This is fixed. I think you must've been looking at an earlier commit (sorry they're not so easy to follow).

---

*[insert.md, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDT9q99P7CyN9zwBKK:-KIDZskD9y-fu_Nnm5gs:-174237034) ([raw file](https://github.com/cockroachdb/docs/blob/b5ad7e5304dc5125b7753cae2007c1fc0a940536/insert.md#L34)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> FWIW, the set expressions are also more flexible
</details>
I didn't realize this. Can you give me some examples?

---

*[insert.md, line 12 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDVCEmIVQBg2HqMRgD:-KIDZ7kGbbYzATgECGgl:1969887102) ([raw file](https://github.com/cockroachdb/docs/blob/90015851270912ca60aa16b57cdc121a7a4f57c9/insert.md#L12)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> UPDATE should only be necessary for ON CONFLICT DO UPDATE. DO NOTHING should still only require INSERT (though I realize this is not currently true, I'll go file an issue: https://github.com/cockroachdb/cockroach/issues/6826)
</details>
Changing `ON CONFLICT` to `ON CONFLICT DO UPDATE` to be more specific.

---

*[upsert.md, line 8 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDUF67ax2csouha3Cr:-KIDpavBQJsVgVWOx-A8:1177214857) ([raw file](https://github.com/cockroachdb/docs/blob/b5ad7e5304dc5125b7753cae2007c1fc0a940536/upsert.md#L8)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Any UPSERT statement can be exactly transformed into an INSERT ... ON CONFLICT. I'd love to see this transformation called out.
> 
> `UPSERT INTO foo (a, b) VALUES (1, 2)` *exactly* transforms into `INSERT INTO foo (a, b) VALUES (1, 2) ON CONFLICT (<primary key columns>) DO UPDATE SET b = excluded b` where there is a SET expr for every column that's in qualified_name list but not in on_conflict
</details>
Adding a "How `UPSERT` Tranforms into `INSERT ON CONFLICT`" section.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/308)*

**Comment by paperstreet** [comment]
:lgtm:

<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Great stuff! I've never reviewed public facing docs before, so I erred on the side of overcommunication. Feel free to ignore any suggestions you feel aren't really appropriate or are too pedantic
</details>


---

Review status: 0 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[insert.md, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDT9q99P7CyN9zwBKK:-KIEKX7iglTPxhLc9rkA:-109564244) ([raw file](https://github.com/cockroachdb/docs/blob/b5ad7e5304dc5125b7753cae2007c1fc0a940536/insert.md#L34)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> I didn't realize this. Can you give me some examples?
</details>
It accepts any expression that UPDATE does. A useful concrete example is counting things:

`CREATE TABLE counters (name STRING PRIMARY KEY, count INT); INSERT INTO counters VALUES ('foo', 1) ON CONFLICT (name) DO UPDATE SET count = counters.count + excluded.count;`

---

*[upsert.md, line 14 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIEMUpK7-u7jiuvR9es:-KIEMUpLmBom3YdGMPuW:-710008652) ([raw file](https://github.com/cockroachdb/docs/blob/fc8dbb4a87a7284b038a4c8dda505443c73c9c31/upsert.md#L14)):*
> ```Markdown
> ## Required Privileges
> 
> When there is no uniqueness conflict on the primary key, the user must have the `INSERT` [privilege](privileges.html) on the table. When there is a uniqueness conflict on the primary key, the user must have the `UPDATE` privilege on the table. 
> ```

UPSERT always requires both privileges

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/308#-:-KIEMxEsptKMb_lvsMRr:1417088721)*

**Comment by jseldess** [comment]
Review status: 0 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[insert.md, line 34 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIDT9q99P7CyN9zwBKK:-KIEPHwtEV0HxpLq1n95:1808181452) ([raw file](https://github.com/cockroachdb/docs/blob/b5ad7e5304dc5125b7753cae2007c1fc0a940536/insert.md#L34)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> It accepts any expression that UPDATE does. A useful concrete example is counting things:
> 
> `CREATE TABLE counters (name STRING PRIMARY KEY, count INT); INSERT INTO counters VALUES ('foo', 1) ON CONFLICT (name) DO UPDATE SET count = counters.count + excluded.count;`
</details>
Added reference to `UPDATE`.

---

*[upsert.md, line 14 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/docs/308#-KIEMUpK7-u7jiuvR9es:-KIEPPXgZEYC2eaolL6d:-499146009) ([raw file](https://github.com/cockroachdb/docs/blob/fc8dbb4a87a7284b038a4c8dda505443c73c9c31/upsert.md#L14)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> UPSERT always requires both privileges
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/308)*
